storage: add `TestMVCCHistories` cases

erikgrinaker opened the pull request:
**storage: add `TestMVCCHistories` metamorphic param for peek bounds**

This patch adds a metamorphic test parameter that enables peek bounds
for MVCC range key-related commands. These peek bounds will always
result in identical, correct MVCC stats.

Unfortunately, we can't assert that the peek bounds are enforced,
because importing the `spanset` package causes a cyclic dependency.

Release note: None
  
**storage: use `del_range_ts` in `TestMVCCHistories`**

Many of the early `TestMVCCHistories` tests for MVCC range tombstones
used the direct engine method `put_rangekey` rather than the MVCC range
tombstone write `del_range_ts` (which does conflict checks, MVCC stats
adjustments, etc), because the latter did not exist yet.

This patch migrates most tests to `del_range_ts`, and also renames some
of the test files to refer to range tombstone rather than range key.
Stats assertions have also been enabled for some tests.

There are no significant changes to the tests themselves.

Release note: None
  
**storage: add `TestMVCCHistories` cases**

This patch adds additional test cases for `TestMVCCHistories`. These
were primarily designed by deliberately introducing bugs in MVCC code
that did not cause existing test cases to fail.

Resolves #86655.

Release note: None

erikgrinaker committed:
storage: use `del_range_ts` in `TestMVCCHistories`

Many of the early `TestMVCCHistories` tests for MVCC range tombstones
used the direct engine method `put_rangekey` rather than the MVCC range
tombstone write `del_range_ts` (which does conflict checks, MVCC stats
adjustments, etc), because the latter did not exist yet.

This patch migrates most tests to `del_range_ts`, and also renames some
of the test files to refer to range tombstone rather than range key.
Stats assertions have also been enabled for some tests.

There are no significant changes to the tests themselves.

Release note: None

erikgrinaker committed:
storage: add `TestMVCCHistories` cases

This patch adds additional test cases for `TestMVCCHistories`. These
were primarily designed by deliberately introducing bugs in MVCC code
that did not cause existing test cases to fail.

Release note: None

erikgrinaker committed:
storage: add `TestMVCCHistories` metamorphic param for peek bounds

This patch adds a metamorphic test parameter that enables peek bounds
for MVCC range key-related commands, and asserts that command accesses
respect the bounds. These peek bounds will always result in identical,
correct MVCC stats.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89004)
<!-- Reviewable:end -->

tbg reviewed (approved):
> Reviewed 1 of 1 files at r1, 13 of 13 files at r2, 10 of 10 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89004)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89004#-NDC1pP812barsi-aP4b):*
https://github.com/cockroachdb/cockroach/pull/80636 might address this, I had to fix a similar problem there. Maybe it can be pulled in here without too much trouble?
___
*[`-- commits` line 34 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89004#-NDC2NNIAYCz8ALN950K):*
:brain: 


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89004)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89004#-NDC1pP812barsi-aP4b):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/pull/80636 might address this, I had to fix a similar problem there. Maybe it can be pulled in here without too much trouble?
</blockquote></details>

I don't think that will be sufficient -- it only allows use of `DisableSpansetAssertions`, not the spanset assertions themselves.

The real fix here is to pull `TestMVCCHistories` out into `storage_test`, so that it can make use of both `spanset` and `storage` without any inbound imports. It doesn't actually look like that would be terrible, since I don't see references to any internal code here. I'll give it a shot.


<!-- Sent from Reviewable.io -->

erikgrinaker committed:
storage: move `TestMVCCHistories` to `storage_test`

This patch moves `TestMVCCHistories` into a separate `storage_test`
package. This will allow it to make use of the `spanset` package for
spanset assertions in a future patch.

This required exporting a handful of private `storage` package members.
There are no functional changes.

Release note: None

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89004)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rhu713 and @tbg)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89004#-NDC1pP812barsi-aP4b):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I don't think that will be sufficient -- it only allows use of `DisableSpansetAssertions`, not the spanset assertions themselves.

The real fix here is to pull `TestMVCCHistories` out into `storage_test`, so that it can make use of both `spanset` and `storage` without any inbound imports. It doesn't actually look like that would be terrible, since I don't see references to any internal code here. I'll give it a shot.
</blockquote></details>

Done. I had to pull in #89026 as well, will merge that first.


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
CI failures are flakes.

bors r+
